Source/Particles/ParticleBoundaryBuffer.cpp

```diff
@@ -0,0 +1,242 @@
+/* Copyright 2021 Andrew Myers
+ *
+ * This file is part of WarpX.
+ *
+ * License: BSD-3-Clause-LBNL
+ */
+
+
+#include "EmbeddedBoundary/DistanceToEB.H"
+#include "Particles/ParticleBoundaryBuffer.H"
+#include "Particles/MultiParticleContainer.H"
+
+#include <AMReX_Geometry.H>
+#include <AMReX_ParmParse.H>
+
+struct IsOutsideDomainBoundary {
+    amrex::GpuArray<amrex::Real, AMREX_SPACEDIM> m_plo;
+    amrex::GpuArray<amrex::Real, AMREX_SPACEDIM> m_phi;
+    int m_idim;
+    int m_iside;
+
+    template <typename SrcData>
+    AMREX_GPU_DEVICE AMREX_FORCE_INLINE
+    int operator() (const SrcData& src,
+                    int ip, const amrex::RandomEngine& /*engine*/) const noexcept
+    {
+        const auto& p = src.getSuperParticle(ip);
+        if (m_iside == 0) {
+            if (p.pos(m_idim) < m_plo[m_idim]) { return 1; }
+        } else {
+            if (p.pos(m_idim) >= m_phi[m_idim]) { return 1; }
+        }
+        return 0;
+    }
+};
+
+struct CopyAndTimestamp {
+    int m_index;
+    int m_step;
+
+    template <typename DstData, typename SrcData>
+    AMREX_GPU_HOST_DEVICE
+    void operator() (const DstData& dst, const SrcData& src,
+                     int src_i, int dst_i) const noexcept
+    {
+        dst.m_aos[dst_i] = src.m_aos[src_i];
+        for (int j = 0; j < SrcData::NAR; ++j)
+            dst.m_rdata[j][dst_i] = src.m_rdata[j][src_i];
+        for (int j = 0; j < src.m_num_runtime_real; ++j)
+            dst.m_runtime_rdata[j][dst_i] = src.m_runtime_rdata[j][src_i];
+        for (int j = 0; j < src.m_num_runtime_int; ++j)
+            dst.m_runtime_idata[j][dst_i] = src.m_runtime_idata[j][src_i];
+        dst.m_runtime_idata[m_index][dst_i] = m_step;
+    }
+};
+
+ParticleBoundaryBuffer::ParticleBoundaryBuffer ()
+{
+    m_particle_containers.resize(numBoundaries());
+
+    for (int i = 0; i < numBoundaries(); ++i)
+    {
+        m_particle_containers[i].resize(numSpecies());
+    }
+
+    m_do_boundary_buffer.resize(numBoundaries());
+    for (int i = 0; i < numBoundaries(); ++i)
+    {
+        m_do_boundary_buffer[i].resize(numSpecies(), 0);
+    }
+
+    for (int i = 0; i < numBoundaries(); ++i)
+    {
+        for (int ispecies = 0; ispecies < numSpecies(); ++ispecies)
+        {
+            amrex::ParmParse pp_species(getSpeciesNames()[ispecies]);
+            pp_species.query("save_particles_at_xlo", m_do_boundary_buffer[0][ispecies]);
```

Comment:
We need to document these new options in `Docs/source/usage/parameters.rst`, I think

Reply:
Discussed on Slack: we'll add those in #2164